From a8db401961a1e003e592a3696902ed6b70167ebd Mon Sep 17 00:00:00 2001
From: Dominique Belhachemi <db@domibel.de>
Date: Fri, 1 May 2026 15:52:57 -0400
Subject: [PATCH 28/37] drm/imagination: flush pending fence release work
 before module teardown

(cherry picked from commit 0aa7103506bee45d96f6c898ef99c1ff8e53116e)
Upstream-Status: Pending [from https://github.com/domibel/linux branch powervr_on_jh7110_visionfive2_v7.3-rc5]

---
 drivers/gpu/drm/imagination/pvr_drv.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/gpu/drm/imagination/pvr_drv.c b/drivers/gpu/drm/imagination/pvr_drv.c
index 58087e8517d6..a0e9a4729570 100644
--- a/drivers/gpu/drm/imagination/pvr_drv.c
+++ b/drivers/gpu/drm/imagination/pvr_drv.c
@@ -1475,6 +1475,10 @@ static void pvr_remove(struct platform_device *plat_dev)
 	xa_destroy(&pvr_dev->free_list_ids);
 
 	pm_runtime_suspend(drm_dev->dev);
+
+	if (pvr_dev->sched_wq)
+		flush_workqueue(pvr_dev->sched_wq);
+
 	pvr_device_fini(pvr_dev);
 	drm_dev_unplug(drm_dev);
 	pvr_watchdog_fini(pvr_dev);
